Clean up the connector's flow control, fix a crash

LukeShu opened the pull request:
## Description

99% of this is trying to clean up the connector's flow control so that I could reason about it to figure out how to fix the race condition that https://github.com/telepresenceio/telepresence/pull/1857 was hitting.  I had thought that it was a pre-existing race condition that this PR would be fixing, but instead since the race was introduced in #1857 I went ahead and fixed it in #1857.  But I was still left with a bunch of cleanup that is probably worth landing.

And while I was looking at all of the connector code to clean it up, I noticed a possible crash during shutdown.  So go ahead and fix that.

The connector has a bunch of objects for its worker goroutines, and they're all in the same package and so can reach in to eachothers struct members directly.  It makes it really hard to reason about when this or that gets set.  So, the theme of this PR is to move each of them in to its own package, and reduce the surface area where they interact.  So the PR is actually pretty small, but the diff is big because of a lot of renaming files and changing whether things are public/private.

## Checklist

<!--
  Please review the requirements for each checkbox, and check them
  off (change "[ ]" to "[x]") as you verify that they are complete.
-->

 - [x] I made sure to update `./CHANGELOG.md`. - yes
 - [x] I made sure to either submit a docs PR, or tell Matt about the necessary documentation changes. - no docs changes besides release notes, which will get synced from changelog
 - [x] My change is adequately tested. - yes
 - [x] I updated `DEVELOPING.md` with any any special dev tricks I had to use to work on this code efficiently. - no tricks

LukeShu committed:
connector: Add more comments

Signed-off-by: Luke Shumaker <lukeshu@datawire.io>

LukeShu committed:
connector: Simplify control flow a bit

Signed-off-by: Luke Shumaker <lukeshu@datawire.io>

LukeShu committed:
connector: Pull the gRPC server in to a separate package

This is stab 1 at having less visible shared state.

Signed-off-by: Luke Shumaker <lukeshu@datawire.io>

LukeShu committed:
connector: Pull the Kubernetes stuff in to a separate package

This is stab #2 at cutting down the large shared state.

Signed-off-by: Luke Shumaker <lukeshu@datawire.io>

LukeShu committed:
connector: userd_k8s: Do a better job of hiding implementation details

Signed-off-by: Luke Shumaker <lukeshu@datawire.io>

LukeShu committed:
connector: Move the Cluster to be a member of the sharedState object

Signed-off-by: Luke Shumaker <lukeshu@datawire.io>

LukeShu committed:
connector: Move the traffic manager stuff in to a separate package

Signed-off-by: Luke Shumaker <lukeshu@datawire.io>

LukeShu committed:
main.mk: Touch up whitespace

Signed-off-by: Luke Shumaker <lukeshu@datawire.io>

LukeShu committed:
connector: Fix potential crash during shutdown

Signed-off-by: Luke Shumaker <lukeshu@datawire.io>

LukeShu committed:
cd pkg/client/connector && mkdir userd_trafficmgr && mv testdata userd_trafficmgr # [ci-skip]

This is to cut down on noise in the next commit to keep the patch
readable.

Signed-off-by: Luke Shumaker <lukeshu@datawire.io>

thallgren commented on pkg/client/connector/command.go:
```diff
@@ -286,9 +280,9 @@ func (s *service) connectWorker(c context.Context, cr *rpc.ConnectRequest, k8sCo
-		ClusterId:      s.cluster.GetClusterId(c),
+		ClusterContext: s.sharedState.GetClusterNonBlocking().K8sConfig.Context,
+		ClusterServer:  s.sharedState.GetClusterNonBlocking().K8sConfig.Server,
+		ClusterId:      s.sharedState.GetClusterNonBlocking().GetClusterId(c),
```
Please assign the result of `GetClusterNonBlocking()` to a variable at line 221 instead of calling it repeatedly 6 times.

thallgren commented on pkg/client/connector/command.go:
```diff
@@ -218,11 +219,11 @@ func (s *service) connectWorker(c context.Context, cr *rpc.ConnectRequest, k8sCo
 
 	dlog.Info(c, "Connecting to traffic manager...")
-	tmgr, err := newTrafficManager(c,
+	tmgr, err := worker_trafficmanager.NewTrafficManager(c,
```
The naming of the package and the constructor seems a bit off. Instead of repeated names in package and constructor, how about `trafficmanager.NewWorker()` ? If we want to have `worker` in the package structure, then perhaps use `pkg/client/connector/worker/trafficmanager`, `pkg/client/connector/worker/grpc`, and `client/connector/worker/cluster`.

thallgren reviewed (changes requested):
Looks good overall. I like the split and encapsulation. I have two concerns:
1. Unnecessary repeated calls of a function should be avoided.
2. The names of packages and constructors can be improved.

LukeShu committed:
telepresence_test.go: Bump one of the timeouts

This was regularly tripping on my laptop. :shrug:

Signed-off-by: Luke Shumaker <lukeshu@datawire.io>

LukeShu committed:
mv pkg/client/connector/auth pkg/client/connector/userd_auth

This is to make it consistent with new package names that will be added in
later commits.

Signed-off-by: Luke Shumaker <lukeshu@datawire.io>

LukeShu committed:
telepresence_test.go: Bump some more timeouts

I see these fail in CI sometimes.

Signed-off-by: Luke Shumaker <lukeshu@datawire.io>

thallgren reviewed (approved):
All good.

LukeShu committed:
Merge remote-tracking branch 'origin/release/v2' into lukeshu/connector-cleanup

Signed-off-by: Luke Shumaker <lukeshu@datawire.io>

LukeShu commented:
Resolved logical merge conflicts. No other changes. Will merge once CI turns green.
